### PR TITLE
fix(isochrones): correct and unify edge splitting

### PR DESCRIPTION
**Pull request by aoles**
### Pull Request Checklist

- [x] 1. I have [**rebased**][rebase] the latest version of the main branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [ ] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [x] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [x] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [x] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [ ] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

### Information about the changes
- Key functionality added: Split edges based on difference in coordinates rather than their actual distance in meters, and use the same splitting length across isochrone builders.
- Reason for change: Distance-based splitting proved to be unreliable, while different defaults across the builders introduced additional differences in the generated geometries.

In order to guarantee correct construction of the concave hull edge splitting operations need to be performed based on the differences computed on geocoordinates rather than on the actual distances in meters. In practice, this switch makes  little difference to the produced geometries and has virtually no measurable impact on performance. Typical differences in geometries observed are as follows, where blue is the geometry produced with  the previous approach, and red with the new one.

image

The reduction of the splitting length has more noticeable effects, see [report](https://rpubs.com/aoles/splitting_length_reduction). For regular isochrones going down by 1/3 from `0.012` to `0.009`, which corresponds to 1333 and 999 meters in latitude, respectively, comes with a performance penalty of around 15%. However, it helps to increase isochrone accuracy and avoid false inclusion of unreachable areas, which is best illustrated by the following example.

image
<sup>Location 10</sup>

Typically, the observed differences are more subtle, such as

image
<sup>Location 4</sup>

and

image
<sup>Location 7</sup>

[config]: https://GIScience.github.io/openrouteservice/run-instance/configuration/
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/main/CONTRIBUTE.md#pull-request-guidelines

### CODE REVIEW

**Comment by sfendrich** [comment]
Could you please elaborate on how this change affects isolines across the globe? Using geo-coordinates means measuring in degrees. The distance between coordinates varies a lot depending on the location and the used coordinate system. Imagine you would invent a coordinate system having its north pole in Berlin. would the isolines still have the same shape?

File: ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java [hunk 1]
```diff
@@ -22,11 +22,11 @@
 import static org.locationtech.jts.algorithm.hull.ConcaveHull.concaveHullByLength;
 
 public abstract class AbstractIsochroneMapBuilder implements IsochroneMapBuilder {
-    private static final double MAX_SPLIT_LENGTH = 20000.0;
+    private static final double MAX_SEGMENT_LENGTH = 0.18;// corresponds to 20000 m
```
Comment (sfendrich) [inline ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java]: What is the measurement unit?

File: ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java [hunk 2]
```diff
@@ -141,12 +141,16 @@ protected void addPointsFromEdge(List<Coordinate> points, double isolineCost, fl
      * @param maxlim limit above which the edge will NOT be split anymore (in meters)
      */
     protected void splitLineSegment(double lat0, double lon0, double lat1, double lon1, List<Coordinate> points, double minlim, double maxlim) {
-        double dist = dcFast.calcDist(lat0, lon0, lat1, lon1);
+        double dx = (lon1 - lon0);
+        double dy = (lat1 - lat0);
+        double normLength = Math.sqrt((dx * dx) + (dy * dy));
```
Comment (sfendrich) [inline ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java:146]: Why can we use cartesian norm in a non-cartesian coordinate system?

**Comment by aoles** [comment]
Thanks @sfendrich, that's a fair point! What this PR tries to achieve is to perform edge splitting in the same coordinate system in which the concave hull is also being computed. The concave hull algorithm is parameterized by the target maximum edge length. Edge splitting makes sure that edge segments which are input to the concave hull algorithm do not exceed this limit.  This PR does not make any change to the concave hull part which actually computes the isochrone polygons. This part has alwas been done in the geocoordinate domain. Having said this, I agree that the generated shape is influenced by the location.

**Comment by sfendrich** [comment]
So, if I build the same town two times, once in Alaska and once in Cameroon, the isochrones would differ in their shape? This sounds undesirable.

**Comment by aoles** [comment]
While the above is true, this is also how it has always been like. Let me stress this again: this PR does not change how the isochrones are build in general. It just makes adjustments to the logic how the edges are split, and the splitting is there only to facilitate correct results from computing the concave hull.

Having said this, I don't think this discussion should be continued here. If you are not satisfied with the current solution of generating isochrone geometries you are welcome to open an issue. Cheers!

**Comment by rtroilo** [comment]
> While the above is true, this is also how it has always been like. Let me stress this again: this PR does not change how the isochrones are build in general. It just makes adjustments to the logic how the edges are split, and the splitting is there only to facilitate correct results from computing the concave hull.
> 
> Having said this, I don't think this discussion should be continued here. If you are not satisfied with the current solution of generating isochrone geometries you are welcome to open an issue. Cheers!

But I think this could have a very easy fix, just make the smoothing-factor depending an the latitude of the center. What do you think?

**Comment by aoles** [comment]
@rtroilo thanks for chiming in. I've thought about this too, however, the further apart you move from the equator the higher  the discrepancy between the length in meters of one latitudinal and one longitudinal degrees. The algorithm which computes the concave hull is parametrized only with one length and does not differentiate between XY directions. So effectively you still have the same issue, or do I miss something? 🤔

**Review by sfendrich — CHANGES_REQUESTED** [review]
Just two minor comments.

**Comment by sfendrich** [comment]
So, to summarize the discussion:  

**The issue solved by this PR** is the following: If a long graph edge is used in an isochrone it should be contained in the concave hull. However, due to the length of the edge, the concave hull algorithm might inappropriately introduce a cave such that the edge is outside the hull. Therefore, we introduce additional nodes on this long edge to split it into multiple shorter ones, yielding more points to be fed into the concave hull algorithm. These additional points prevent the creation of the cave.

**Independent of this PR** it seems there is something fundamentally wrong with isochrones being computed from geo-coordinates. At the scale of a single isochrone, this is probably not relevant. But when comparing isochrones at different latitudes, the length differences of a longitude degree might affect the shape and size of the isochrone's geometry.
Here is a new issue to continue the discussion on this second topic: https://github.com/GIScience/openrouteservice/issues/1711.

File: ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java (hunk 1 above)
Comment (aoles) [reply ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java]: Degrees.

File: ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java (hunk 2 above)
Comment (aoles) [reply ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java:146]: This is based on the fact that the rest of polygon processing, in particular the concave hull algorithm, assumes cartesian coordinates. I believe for smaller distances such approximation is legitimate. In fact, this is not the only place such norm is being used. Similar computations are being performed when adding buffer points,

https://github.com/GIScience/openrouteservice/blob/a890ad2230e8594b0000a35e687683641ef5c5ba/ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java#L78-L82

 and in the contour preprocessing part.

https://github.com/GIScience/openrouteservice/blob/a890ad2230e8594b0000a35e687683641ef5c5ba/ors-engine/src/main/java/org/heigit/ors/fastisochrones/Contour.java#L385-L388

File: ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java (hunk 1 above)
Comment (sfendrich) [reply ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java]: Then I suggest renaming it to `MAX_SEGMENT_ANGLE`.

**Review by sfendrich — APPROVED** [review]

File: ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java (hunk 1 above)
Comment (sfendrich) [reply ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java]: I've renamed it.

Comment (aoles) [reply ors-engine/src/main/java/org/heigit/ors/isochrones/builders/AbstractIsochroneMapBuilder.java]: Thank you for going forward and let me apologize for not responding to you earlier, I was busy with other stuff :innocent:

**Comment by MichaelsJP** [comment]
@aoles Can we merge this?

**Pull request merged**